File: modules/wordfold-pronunciation/android/src/main/java/expo/modules/wordfoldpronunciation/WordfoldPronunciationModule.kt
package expo.modules.wordfoldpronunciation

import android.net.Uri
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.types.OptimizedRecord
import java.io.File
import java.util.UUID

class WordfoldPronunciationModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("WordfoldPronunciation")

    AsyncFunction("synthesizeToFile") { options: PronunciationFileOptions, promise: Promise ->
      val context = appContext.reactContext
      if (context == null) {
        promise.reject("E_CONTEXT", "The Android application context is unavailable", null)
        return@AsyncFunction
      }
      if (options.text.isBlank()) {
        promise.reject("E_INPUT", "Pronunciation text cannot be empty", null)
        return@AsyncFunction
      }
      if (options.text.length > TextToSpeech.getMaxSpeechInputLength()) {
        promise.reject("E_INPUT", "Pronunciation text is too long for the device speech engine", null)
        return@AsyncFunction
      }

      val outputFile = fileForUri(options.outputUri)
      if (outputFile == null) {
        promise.reject("E_OUTPUT", "Pronunciation output must use a local file URI", null)
        return@AsyncFunction
      }
      outputFile.parentFile?.mkdirs()
      outputFile.delete()

      val utteranceId = UUID.randomUUID().toString()
      var engine: TextToSpeech? = null
      var settled = false
      fun reject(code: String, message: String, error: Throwable? = null) {
        if (settled) return
        settled = true
        outputFile.delete()
        promise.reject(code, message, error)
        engine?.shutdown()
      }
      fun resolve() {
        if (settled) return
        val size = outputFile.length()
        if (!outputFile.isFile || size <= MINIMUM_AUDIO_FILE_BYTES) {
          reject("E_EMPTY_AUDIO", "The device speech engine produced no playable audio")
          return
        }
        settled = true
        promise.resolve(size.toDouble())
        engine?.shutdown()
      }

      engine = TextToSpeech(context) { status ->
        val initializedEngine = engine
        if (status != TextToSpeech.SUCCESS || initializedEngine == null) {
          reject("E_TTS_INIT", "The device speech engine could not be initialized")
          return@TextToSpeech
        }
        val voice = try {
          initializedEngine.voices.firstOrNull { it.name == options.voiceIdentifier }
        } catch (error: Exception) {
          reject("E_VOICE", "The installed device voices could not be read", error)
          return@TextToSpeech
        }
        if (voice == null) {
          reject("E_VOICE", "The exact requested device voice is no longer installed")
          return@TextToSpeech
        }
        if (canonicalLocale(voice.locale.toLanguageTag()) != canonicalLocale(options.locale)) {
          reject("E_VOICE_LOCALE", "The requested device voice no longer matches the exact pronunciation locale")
          return@TextToSpeech
        }

        initializedEngine.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
          override fun onStart(id: String) = Unit
          override fun onDone(id: String) {
            if (id == utteranceId) resolve()
          }
          override fun onStop(id: String, interrupted: Boolean) {
            if (id == utteranceId) reject("E_TTS_STOPPED", "Pronunciation file synthesis was stopped")
          }
          override fun onError(id: String) {
            if (id == utteranceId) reject("E_TTS", "The device speech engine could not synthesize pronunciation")
          }
          override fun onError(id: String, errorCode: Int) {
            if (id == utteranceId) reject("E_TTS", "The device speech engine failed with code $errorCode")
          }
        })
        initializedEngine.voice = voice
        initializedEngine.setSpeechRate(options.rate)
        initializedEngine.setPitch(options.pitch)
        val params = Bundle().apply {
          putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, 1f)
        }
        val result = initializedEngine.synthesizeToFile(options.text, params, outputFile, utteranceId)
        if (result == TextToSpeech.ERROR) {
          reject("E_TTS", "The device speech engine rejected pronunciation file synthesis")
        }
      }
    }
  }

  private fun fileForUri(uri: String): File? {
    val parsed = Uri.parse(uri)
    if (parsed.scheme != "file" || parsed.path.isNullOrBlank()) return null
    return File(parsed.path!!)
  }

  private fun canonicalLocale(locale: String) = locale.trim().replace('_', '-').lowercase()

  companion object {
    private const val MINIMUM_AUDIO_FILE_BYTES = 44L
  }
}

@OptimizedRecord
data class PronunciationFileOptions(
  @Field val text: String,
  @Field val locale: String,
  @Field val voiceIdentifier: String,
  @Field val rate: Float,
  @Field val pitch: Float,
  @Field val outputUri: String,
) : Record

File: modules/wordfold-pronunciation/ios/WordfoldPronunciationModule.swift
import AVFoundation
import ExpoModulesCore

public final class WordfoldPronunciationModule: Module {
  private var tasks: [UUID: PronunciationFileTask] = [:]

  public func definition() -> ModuleDefinition {
    Name("WordfoldPronunciation")

    AsyncFunction("synthesizeToFile") { (options: PronunciationFileOptions, promise: Promise) in
      guard !options.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        promise.reject("E_INPUT", "Pronunciation text cannot be empty")
        return
      }
      guard let voice = AVSpeechSynthesisVoice(identifier: options.voiceIdentifier) else {
        promise.reject("E_VOICE", "The exact requested device voice is no longer installed")
        return
      }
      guard canonicalLocale(voice.language) == canonicalLocale(options.locale) else {
        promise.reject("E_VOICE_LOCALE", "The requested device voice no longer matches the exact pronunciation locale")
        return
      }
      guard let outputURL = URL(string: options.outputUri), outputURL.isFileURL else {
        promise.reject("E_OUTPUT", "Pronunciation output must use a local file URI")
        return
      }

      let id = UUID()
      let task = PronunciationFileTask(
        text: options.text,
        voice: voice,
        rate: options.rate,
        pitch: options.pitch,
        outputURL: outputURL,
        promise: promise
      ) { [weak self] in
        self?.tasks[id] = nil
      }
      tasks[id] = task
      task.start()
    }

    OnDestroy {
      let activeTasks = Array(self.tasks.values)
      self.tasks.removeAll()
      activeTasks.forEach { $0.cancel() }
    }
  }

  private func canonicalLocale(_ locale: String) -> String {
    locale.trimmingCharacters(in: .whitespacesAndNewlines)
      .replacingOccurrences(of: "_", with: "-")
      .lowercased()
  }
}

private struct PronunciationFileOptions: Record {
  @Field var text: String = ""
  @Field var locale: String = ""
  @Field var voiceIdentifier: String = ""
  @Field var rate: Double = 0.9
  @Field var pitch: Double = 1
  @Field var outputUri: String = ""
}

private final class PronunciationFileTask {
  private let synthesizer = AVSpeechSynthesizer()
  private let utterance: AVSpeechUtterance
  private let outputURL: URL
  private let promise: Promise
  private let finish: () -> Void
  private var audioFile: AVAudioFile?
  private var settled = false

  init(
    text: String,
    voice: AVSpeechSynthesisVoice,
    rate: Double,
    pitch: Double,
    outputURL: URL,
    promise: Promise,
    finish: @escaping () -> Void
  ) {
    let utterance = AVSpeechUtterance(string: text)
    utterance.voice = voice
    utterance.rate = Float(rate) * AVSpeechUtteranceDefaultSpeechRate
    utterance.pitchMultiplier = Float(pitch)
    utterance.volume = 1
    self.utterance = utterance
    self.outputURL = outputURL
    self.promise = promise
    self.finish = finish
  }

  func start() {
    do {
      try FileManager.default.createDirectory(
        at: outputURL.deletingLastPathComponent(),
        withIntermediateDirectories: true
      )
      try? FileManager.default.removeItem(at: outputURL)
    } catch {
      reject("E_OUTPUT", "The pronunciation cache directory could not be created", error)
      return
    }

    synthesizer.write(utterance) { [weak self] buffer in
      self?.receive(buffer)
    }
  }

  func cancel() {
    guard !settled else { return }
    synthesizer.stopSpeaking(at: .immediate)
    reject("E_TTS_STOPPED", "Pronunciation file synthesis was stopped")
  }

  private func receive(_ buffer: AVAudioBuffer) {
    guard !settled else { return }
    guard let pcmBuffer = buffer as? AVAudioPCMBuffer else {
      reject("E_AUDIO_BUFFER", "The device speech engine returned an unsupported audio buffer")
      return
    }
    if pcmBuffer.frameLength == 0 {
      audioFile = nil
      resolveFile()
      return
    }
    do {
      if audioFile == nil {
        audioFile = try AVAudioFile(
          forWriting: outputURL,
          settings: pcmBuffer.format.settings,
          commonFormat: pcmBuffer.format.commonFormat,
          interleaved: pcmBuffer.format.isInterleaved
        )
      }
      try audioFile?.write(from: pcmBuffer)
    } catch {
      reject("E_AUDIO_WRITE", "The synthesized pronunciation could not be written", error)
    }
  }

  private func resolveFile() {
    guard !settled else { return }
    do {
      let attributes = try FileManager.default.attributesOfItem(atPath: outputURL.path)
      let size = (attributes[.size] as? NSNumber)?.intValue ?? 0
      guard size > 44 else {
        reject("E_EMPTY_AUDIO", "The device speech engine produced no playable audio")
        return
      }
      settled = true
      promise.resolve(size)
      finish()
    } catch {
      reject("E_EMPTY_AUDIO", "The device speech engine produced no playable audio", error)
    }
  }

  private func reject(_ code: String, _ message: String, _ error: Error? = nil) {
    guard !settled else { return }
    settled = true
    audioFile = nil
    try? FileManager.default.removeItem(at: outputURL)
    let description = error.map { "\(message): \($0.localizedDescription)" } ?? message
    promise.reject(code, description)
    finish()
  }
}
